File: app/src/main/java/com/mohsenoid/rickandmorty/view/character/details/CharacterDetailsPresenter.java
package com.mohsenoid.rickandmorty.view.character.details;

import com.mohsenoid.rickandmorty.data.DataCallback;
import com.mohsenoid.rickandmorty.data.exception.NoOfflineDataException;
import com.mohsenoid.rickandmorty.domain.Repository;
import com.mohsenoid.rickandmorty.domain.entity.CharacterEntity;
import com.mohsenoid.rickandmorty.util.config.ConfigProvider;


public class CharacterDetailsPresenter implements CharacterDetailsContract.Presenter {

    private final Repository repository;
    private final ConfigProvider configProvider;

    private int characterId;

    private CharacterDetailsContract.View view = null;

    public CharacterDetailsPresenter(Repository repository, ConfigProvider configProvider) {
        this.repository = repository;
        this.configProvider = configProvider;
    }

    @Override
    public void bind(CharacterDetailsContract.View view) {
        this.view = view;
    }

    @Override
    public void unbind() {
        view = null;
    }

    @Override
    public void setCharacterId(int characterId) {
        this.characterId = characterId;
    }

    @Override
    public void loadCharacter() {
        if (view != null) view.showLoading();

        queryCharacter(characterId);
    }

    private void queryCharacter(int characterId) {
        if (!configProvider.isOnline()) {
            if (view != null) view.showOfflineMessage(false);
        }

        repository.queryCharacterDetails(characterId, new DataCallback<CharacterEntity>() {

            @Override
            public void onSuccess(CharacterEntity character) {
                if (view != null) {
                    view.setCharacter(character);
                    view.hideLoading();
                }
            }

            @Override
            public void onError(Exception exception) {
                if (view != null) {
                    view.hideLoading();
                }

                if (exception instanceof NoOfflineDataException) {
                    if (view != null) view.onNoOfflineData();
                } else {
                    if (view != null) view.showMessage(exception.getMessage());
                }
            }
        });
    }
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/view/character/details/CharacterDetailsPresenter.kt
package com.mohsenoid.rickandmorty.view.character.details

import com.mohsenoid.rickandmorty.domain.Repository
import com.mohsenoid.rickandmorty.domain.model.QueryResult
import com.mohsenoid.rickandmorty.util.StatusProvider

class CharacterDetailsPresenter(
    private val repository: Repository,
    private val statusProvider: StatusProvider
) : CharacterDetailsContract.Presenter() {

    private var view: CharacterDetailsContract.View? = null

    override fun bind(view: CharacterDetailsContract.View) {
        this.view = view
    }

    override fun unbind() {
        view = null
    }

    override suspend fun loadCharacter(characterId: Int) {
        view?.showLoading()
        queryCharacter(characterId)
    }

    private suspend fun queryCharacter(characterId: Int) {
        if (!statusProvider.isOnline()) {
            view?.showOfflineMessage(isCritical = false)
        }

        when (val result = repository.getCharacterDetails(characterId)) {
            is QueryResult.Successful -> view?.setCharacter(result.data)
            QueryResult.NoCache -> view?.onNoOfflineData()
            QueryResult.Error -> view?.showMessage("Error fetching character details")
        }

        view?.hideLoading()
    }
}
